src/Compiler/CompileExpr.idr

```diff
@@ -793,13 +794,15 @@ compileDef n
     noDefYet _ _ = False
 
 export
-mkForgetDef : {auto c : Ref Ctxt Defs} ->
-              Name -> Core ()
-mkForgetDef n
+mkForgetDef :  {auto c : Ref Ctxt Defs}
+            -> UsageMap
+            -> Name
+            -> Core ()
+mkForgetDef um n
     = do defs <- get Ctxt
          Just gdef <- lookupCtxtExact n (gamma defs)
               | Nothing => throw (InternalError ("Trying to compile unknown name " ++ show n))
          case compexpr gdef of
               Nothing => pure ()
-              Just cdef => do let ncdef = forgetDef cdef
+              Just cdef => do let ncdef = forgetDef (adjustDef um cdef)
```

Comment:
OK, so I updated the code. However, I am still not sure how to add the newly extracted toplevel definitions to the global context. Right now, I just append them manually when generating the different IRs, but this seems to be problematic when executing functions interactively (for instance, test `node/node011` fails). Any idea how to proceed from here?

Reply:
You could add them as new `GlobalDef`s. I've had a quick look and I don't think it would be that hard to create the defs themselves, but I don't know how you'd go about adding them to the global context.

Reply:
This led to additional issues. I therefore opted for passing around the optimized definitions explicitly without CSE affecting the global context at all. Otherwise I ran into issues in interactive mode, where CSE-optimized definitions where kept in the context without the corresponding new toplevel definitions from lifted subexpressions.